### ISSUE TITLE
Improve how `node_comment_statistics` works

### ISSUE THREAD

**Issue by alanmels** [backdrop/backdrop-issues#5870]
## Description of the bug

Stats update function for node comments does not work if the `node_comment_statistics` table is missing respective records.

## Steps To Reproduce

Encountered this issue when a Views table sorted by `Content: Updated/commented date (desc)` didn't want to display some nodes. The missing nodes came up only when I deleted the filter. That made me realize the `node_comment_statistics` table was missing entries for troublesome nodes. 

DB was never meddled directly, the respective nodes were never deleted, so not sure how the `node_comment_statistics` table could've lost some entries. Maybe the `comment_node_predelete()` was somehow involved. 

Anyway, leaving the problem of how exactly the respective entries in the `node_comment_statistics` table got deleted aside, I believe the `db_update` function responsible for comment statistics must be improved to become fool-prove against situations when respective records somehow are missing in the `node_comment_statistics` table.

You can manually delete an entry in `node_comment_statistics` table remembering `nid` and then try to add comment to that node and see the stats for the node will never update, which in turn makes lots of statistic based Views settings/filters dysfunctional. 

## Actual behavior

The following lines of in https://github.com/backdrop/backdrop/blob/be768da3a31a70bfcc4598bcff0c19132f487070/core/modules/comment/comment.entity.inc#L419

```
db_update('node_comment_statistics')
        ->fields(array(
          'cid' => $last_reply->cid,
          'comment_count' => $count,
          'last_comment_timestamp' => $last_reply->changed,
          'last_comment_name' => $last_reply->uid ? '' : $last_reply->name,
          'last_comment_uid' => $last_reply->uid,
        ))
        ->condition('nid', $nid)
        ->execute();
```

assume the needed `$nid` is always there, which unfortunately is not the case. I suggest to improve the above lines to ensure the respective entry does exist before attempting to update it.

**Comment by alanmels** [comment]
Unfortunately, `db_merge()` is giving the following error:

```
EntityStorageException: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '0' for key 'PRIMARY' in EntityDatabaseStorageController->save() (line 789 of /var/www/html/docroot/core/modules/entity/entity.controller.inc). 
```

**Comment by argiepiano** [comment]
I haven't looked at this issue, but at quick glance, you can't just replace `db_update` with `db_merge`. You have to indicate the key column to compare with the `key` method. See https://www.drupal.org/docs/7/api/database-api/merge-queries-using-db_merge

**Comment by alanmels** [comment]
@argiepiano thanks for taking a look at this. While I'm trying to figure out how to make the `db_merge()` method work, can you elaborate why we can't just replace:

```
      db_update('node_comment_statistics')
        ->fields(array(
          'cid' => $last_reply->cid,
          'comment_count' => $count,
          'last_comment_timestamp' => $last_reply->changed,
          'last_comment_name' => $last_reply->uid ? '' : $last_reply->name,
          'last_comment_uid' => $last_reply->uid,
        ))
        ->condition('nid', $nid)
        ->execute();
```

 with

```
  // Sync with node_comment_statistics table.
      $record = db_query('SELECT * FROM {node_comment_statistics} WHERE nid = :nid', array(
        ':nid' => $nid,
      ))->fetchField();
      if (empty($record)) {
        db_insert('node_comment_statistics')
          ->fields(array(
            'nid' => $nid,
            'cid' => $last_reply->cid,
            'comment_count' => '1',
            'last_comment_timestamp' => $last_reply->changed,
            'last_comment_name' => $last_reply->uid ? '' : $last_reply->name,
            'last_comment_uid' => $last_reply->uid,
          ))
          ->execute();
      }
      else {
        db_update('node_comment_statistics')
          ->fields(array(
            'cid' => $last_reply->cid,
            'comment_count' => $count,
            'last_comment_timestamp' => $last_reply->changed,
            'last_comment_name' => $last_reply->uid ? '' : $last_reply->name,
            'last_comment_uid' => $last_reply->uid,
          ))
          ->condition('nid', $nid)
          ->execute();
      }
```

**Comment by kiamlaluno** [comment]
I think it should be the following code.

```php
db_merge('node_comment_statistics')
  ->fields(array(
    'cid' => $last_reply->cid,
    'comment_count' => $count,
    'last_comment_timestamp' => $last_reply->changed,
    'last_comment_name' => $last_reply->uid ? '' : $last_reply->name,
    'last_comment_uid' => $last_reply->uid,
  ))
  ->key('nid', $nid)
  ->execute();
```

**Comment by alanmels** [comment]
> ```
> db_merge('node_comment_statistics')
>   ->fields(array(
>     'cid' => $last_reply->cid,
>     'comment_count' => $count,
>     'last_comment_timestamp' => $last_reply->changed,
>     'last_comment_name' => $last_reply->uid ? '' : $last_reply->name,
>     'last_comment_uid' => $last_reply->uid,
>   ))
>   ->key('nid', $nid)
>   ->execute();
> ```

Thanks for the suggestion. It's giving:

```
TypeError: Argument 1 passed to MergeQuery::key() must be of the type array, string given, called in /var/www/html/docroot/core/modules/comment/comment.entity.inc on line 426 in MergeQuery->key() (line 1511 of /var/www/html/docroot/core/includes/database/query.inc). 
```

**Comment by alanmels** [comment]
Changing it to `->key('nid', [$nid])` also didn't make an effect.

**Comment by argiepiano** [comment]
Try 

```php
db_merge('node_comment_statistics')
  ->fields(array(
    'cid' => $last_reply->cid,
    'comment_count' => $count,
    'last_comment_timestamp' => $last_reply->changed,
    'last_comment_name' => $last_reply->uid ? '' : $last_reply->name,
    'last_comment_uid' => $last_reply->uid,
  ))
  ->key(array('nid' => $nid))
  ->execute();
```

It'd be helpful if you read the documentation I posted above.

**Comment by alanmels** [comment]
@argiepiano, bingo! Works just as expected. I'm gonna update the PR. Thanks a lot!

**Comment by kiamlaluno** [comment]
It's my fault: I looked at Backdrop code using `db_merge()` but forgot to change the `key()` argument to an array, when I copied the code in the other comment.

Is there any way to reproduce this issue during the normal use of Backdrop? I am asking because randomly deleting rows from the *node_comment_statistics* table is probably not going to be accepted as evidence the code needs to be changed.  
Surely, using `db_merge()` instead of `db_update()` solves any issue when the row for a specific node ID doesn't exist, but we don't change every call to `db_update()` with a call to `db_merge()`.

**Comment by alanmels** [comment]
@kiamlaluno, here i the quote from my initial post:

> DB was never meddled directly, the respective nodes were never deleted, so not sure how the node_comment_statistics table could've lost some entries. Maybe the comment_node_predelete() was somehow involved.

I'm afraid I can't give the exact recipe on how to reproduce the issue, but it happened on live site. It is true the website has lots of custom code, but I searched high and low and could not find anything there that would interfere with the `node_comment_statistics` table. And that was why I thought that maybe comment_node_predelete() was somehow involved.

**Comment by kiamlaluno** [comment]
The function that creates the rows for a node is [`comment_node_insert()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21comment%21comment.module/function/comment_node_insert/1), which uses the following code.

```php
  // Allow bulk updates and inserts to temporarily disable the
  // maintenance of the {node_comment_statistics} table.
  if (state_get('comment_maintain_node_statistics', TRUE)) {
    db_insert('node_comment_statistics')
      ->fields(array(
        'nid' => $node->nid,
        'cid' => 0,
        'last_comment_timestamp' => $node->changed,
        'last_comment_name' => NULL,
        'last_comment_uid' => $node->uid,
        'comment_count' => 0,
      ))
      ->execute();
  }
```

Essentially, it's sufficient `state_get('comment_maintain_node_statistics', TRUE)` returns `FALSE` and the row for that node isn't created. When the comment is then created, `state_get('comment_maintain_node_statistics', TRUE)` could returns `TRUE`; [`CommentStorageController::updateNodeStatistics()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21comment%21comment.entity.inc/function/CommentStorageController%3A%3AupdateNodeStatistics/1) would try to update a row that doesn't exist, with the effect of not writing any data.

This is something that can happen, since the node could be created long before the comment, at the point that `state_get('comment_maintain_node_statistics', TRUE)` returns a different value.

**Comment by kiamlaluno** [comment]
@alanmels I think the steps to reproduce the issue involve a bulk update that temporary disables the maintenance of the *node_comment_statistics* table when the node is created, which is then not disabled when comments for that node are created.  
Doing so avoids that comment statistics are ever kept for that node. I am not sure this is wanted.

**Comment by kiamlaluno** [comment]
As comment for the PR: There are two calls to `db_update('node_comment_statistics')` that need to be replaced with calls to `db_merge('node_comment_statistics')`.

**Comment by argiepiano** [comment]
I also believe this change will not be committed unless there is a way to demonstrate that there is an issue that happens during normal operation of Backdrop.

Two things to consider: 

1. All entries of  the `node_comment_statistics` table for all existing nodes are created when `comment` is enabled, by `comment_enable()`.  
2. There are no instances in core when `state_get('comment_maintain_node_statistics', TRUE);` will return FALSE. Therefore there is no scenario when an entry will need `db_merge()` rather than `db_update()`, since there are entries in that table for all nodes already
3. The only instance I can imagine when this would fail would be when a contrib module somehow sets the `comment_maintain_node_statistics` state to false (for example, as @kiamlaluno pointed out, disabling it temporarily when a contrib module is doing a bulk import or update). I would imagine that such contrib module would then generate all necessary entries in `node_comment_statistics` after the bulk update. If not, that's a bug that needs to be fixed by the contrib module

So, in summary, this fix is for a problem that is most likely generated by a bug outside the function. My suggestion is to figure out why this is happening in the first place. Which is the same as saying that you need to find out a way to reliably reproduce the problem you are trying to fix.

**Comment by argiepiano** [comment]
I can see that the only contrib module to ever disable the state `comment_maintain_node_statistics` temporarily is `migrate` (which BTW is a contrib module that has no release, and likely to be buggy). Therefor, I'm thinking that your issue @alanmels was probably produced by using `migrate` and having it fail or break down while doing stuff?

**Comment by avpaderno** [comment]
I am not contrary to this change, even if I can understand it's not seen as a necessary change. I guess the point is: Is it expected that a call to `state_set('comment_maintain_node_statistics', FALSE)` stops the *node_comment_statistics* table to be filled with information about all the nodes created before `state_set('comment_maintain_node_statistics', TRUE)` is called?

I understand there isn't a bug in Backdrop core, since it doesn't set *comment_maintain_node_statistics* to `FALSE`. On the other side, *comment_maintain_node_statistics* (which contributed modules can set) is described as a flag that allows bulk updates and inserts to temporarily disable the maintenance of the *node_comment_statistics* table, not as a flag that avoids that information about specific nodes is gathered until the Comment module is disabled and re-enabled. If the code isn't changed, at least *comment_maintain_node_statistics* should be better documented so developers know what happens when they set it to `FALSE`, or what they are supposed to do when they change *comment_maintain_node_statistics*.

**Comment by alanmels** [comment]
Thanks for the insights. I agree that at least the documentation should be improved. 

BTW, the newly ported https://github.com/backdrop-contrib/comment_alter module is also involved which has lots to do with comments. That module requires the `diff` module, which is also kind of buggy. Anyway, whatever cause the missing entries, if some less experienced Backdrop user runs onto the problem then s/he will spend lots of sweat before figuring out why some views pages, blocks, filters are not working. So I don't see any reason why `db_insert()` couldn't be replaced with `db_merge()` especially when they acting the same way, but the later is covering more use cases to ensure Backdrop works as expected.

**Comment by izmeez** [comment]
Reading through this issue I am of the same opinion as @klonos that the PR changes are minimal and may be seen as good "defensive coding".

**Comment by izmeez** [comment]
What happened to the PR? I can't find it.

**Comment by avpaderno** [comment]
The PR is https://github.com/backdrop/backdrop/pull/4262.

**Comment by quicksketch** [comment]
Thanks @avpaderno! In order to merge this PR, it needs to be recreated, as the original one is no longer possible to open (or merge).
